crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -155,6 +135,200 @@ impl FormatRule<Expr, PyFormatContext<'_>> for FormatExpr {
+/// comments can become own line comments when we discard nested parentheses.
+///
+/// Style decision: When there are multiple nested parentheses around an expression, we consider the
+/// outermost parentheses the relevant ones and discard the others.
+fn format_with_parentheses_comments(
+    expression: &Expr,
+    node_comments: &LeadingDanglingTrailingComments,
+    f: &mut PyFormatter,
+) -> FormatResult<()> {
+    // First part: Split the comments
+
+    // TODO(konstin): This is copied from `parenthesized_range`, except that we don't have the
+    // parent, which is a problem:
+    // ```python
+    // f(
+    //     # a
+    //     (a)
+    // )
+    // ```
+    // gets formatted as
+    // ```python
+    // f(
+    //     (
+    //         # a
+    //         a
+    //     )
+    // )
+    // ```
+    let right_tokenizer = SimpleTokenizer::starts_at(expression.end(), f.context().source())
+        .skip_trivia()
+        .take_while(|token| token.kind == SimpleTokenKind::RParen);
+
+    let left_tokenizer = BackwardsTokenizer::up_to(
+        expression.start(),
+        f.context().source(),
+        f.context().comments().ranges(),
+    )
+    .skip_trivia()
+    .take_while(|token| token.kind == SimpleTokenKind::LParen);
+
+    // Zip closing parenthesis with opening parenthesis. The order is intentional, as testing for
+    // closing parentheses is cheaper, and `zip` will avoid progressing the `left_tokenizer` if
+    // the `right_tokenizer` is exhausted.
+    let range_with_parens = right_tokenizer
+        .zip(left_tokenizer)
+        .last()
+        .map(|(right, left)| TextRange::new(left.start(), right.end()));
+
+    let (leading_split, trailing_split) = if let Some(range_with_parens) = range_with_parens {
+        let leading_split = node_comments
+            .leading
+            .partition_point(|comment| comment.start() < range_with_parens.start());
+        let trailing_split = node_comments
+            .trailing
+            .partition_point(|comment| comment.start() < range_with_parens.end());
+        (leading_split, trailing_split)
+    } else {
+        (0, node_comments.trailing.len())
+    };
+
+    let leading_outer = &node_comments.leading[..leading_split];
+    let leading_inner = &node_comments.leading[leading_split..];
+    let trailing_inner = &node_comments.trailing[..trailing_split];
+    let trailing_outer = &node_comments.trailing[trailing_split..];
+
+    // Preserve an opening parentheses comment
+    // ```python
+    // a = ( # opening parentheses comment
+    //     # leading inner
+    //     1
+    // )
+    // ```
+    let (parentheses_comment, leading_inner) = match leading_inner.split_first() {
+        Some((first, rest)) if first.line_position().is_end_of_line() => {
+            (slice::from_ref(first), rest)
+        }
+        _ => (Default::default(), node_comments.leading),
+    };
+
+    // Second Part: Format
+
+    // The code order is a bit strange here, we format:
```

Comment:
Can we explain the reason why it has to be strange? Or can we make it not strange?

Reply:
Mainly because we have to construct the nested `format_with`s. I considered inlining everything but then we lack the proper structuring